### ISSUE TITLE
/info/refs http request malformed when base remote path is /

### ISSUE THREAD

**Issue by josharian** [libgit2/libgit2#5321]
### Reproduction steps

Construct an http git remote whose URL has path `/`, such as `http://localhost:8080/`. Do a push.

### Expected behavior

libgit2 sends an HTTP request with URL `/info/refs?service=git-receive-pack`. (This is what command line git does.)

### Actual behavior

libgit2 sends an HTTP request with URL `/info/refs?service=git-receive-pack/info/refs?service=git-receive-pack`.

### Version of libgit2 (release number or SHA1)

7acb764eeeea22cbfaa4dd625bc8516966ed96f3 (master at time of posting)

Remote URLs of this form also fail with v28, but for different reasons. (c6ab183e9c960b74471a7d106a4deb0c8b28a8ec, which postdates v28, fixed other bugs with such URLs.)

### OS

macOS

**Comment by josharian** [comment]
Ah. The duplicate http request line is the result of a 301 being served up by my server.

The initial request being made is for `//info/refs?service=git-receive-pack`. The server sends a 301 to `/info/refs?service=git-receive-pack`. Then the duplicate request is made.

It appears there are two bugs here:

* sending the double leading slash
* appending `/info/refs?service=git-receive-pack` to the response of the 301, rather than just using the url provided in the 301 response

I'm still digging...

**Comment by josharian** [comment]
I suppose it is arguable whether given a https remote url of `http://localhost/`, the request should go to `http://localhost/info/refs?service=git-receive-pack` or to `http://localhost//info/refs?service=git-receive-pack`.

If you want to de-duplicate the `/`, here's a crude patch that does so, at least for experimentation purposes:

```diff
diff --git a/src/transports/http.c b/src/transports/http.c
index 47094f700..b9c8de9c7 100644
--- a/src/transports/http.c
+++ b/src/transports/http.c
@@ -182,8 +182,13 @@ static int gen_request(
        size_t content_length)
 {
        http_subtransport *t = OWNING_SUBTRANSPORT(s);
-       const char *path = t->server.url.path ? t->server.url.path : "/";
+       char *path = t->server.url.path ? t->server.url.path : "/";
        size_t i;
+       size_t pathlen = strlen(path);
+       if (pathlen > 0 && path[pathlen - 1] == '/')
+       {
+               path[pathlen - 1] = 0;
+       }
 
        if (t->proxy_opts.type == GIT_PROXY_SPECIFIED)
                git_buf_printf(buf, "%s %s://%s:%s%s%s HTTP/1.1\r\n",
```

**Comment by josharian** [comment]
I'm out of time for now. :(

On the question of the handling of the 301, here's what I know so far. The 301 triggers a call to `gitno_connection_data_handle_redirect`. At the end of that function is a check to attempt to remove the `service_suffix` (in this case, `/info/refs?service=git-receive-pack`), if present. This is necessary because gen_request will later re-append the service suffix.

I'd say the design of this is flawed: There's no guarantee that 301 reply preserves the original url structure, and if it doesn't, then it would be incorrect to re-append the service suffix. If the server indicated where the resource is, that should be used verbatim. The fix would be not to append the service suffix after 1 or more redirects.

Setting aside the design question, the removal of the service suffix as written does not function correctly. There appears to be confusion in the string index math over whether the query parameters are part of the url path. As a result, the `git__strncmp` finds the suffix incorrectly not present, so it doesn't get stripped, leading to the double-appended result I saw initially.

One last comment about the code as written. It contains:

```c
				/* Ensure we leave a minimum of '/' as the path */
				if (suffix_offset == 0)
					suffix_offset++;
```

If that is to stay, then the removal of the leading / when constructing the service url I hinted at in a previous comment is required. Otherwise we will have reconstructed the exact original url, which will result in another identical 301.

**Comment by josharian** [comment]
cc @ethomson; sorry about all the words

**Comment by ethomson** [comment]
> I'd say the design of this is flawed: There's no guarantee that 301 reply preserves the original url structure, and if it doesn't, then it would be incorrect to re-append the service suffix. If the server indicated where the resource is, that should be used verbatim. The fix would be not to append the service suffix after 1 or more redirects.

The URL structure needs to be preserved.  If you try to hit `http://github.com/foo/bar/info/refs`, and you get redirected to **`https://github.com/foo/bar/info/refs`**, then you need to use that redirect to find the other endpoints for upload-pack and receive-pack.

There must be a correspondence between `/info/refs` paths and upload-pack and receive-pack paths - this is part of the spec.

So we need to unwind the redirects to know where our other endpoints are.  I concede this is a little messy but that's where we are.  But I agree that these:

* sending the double leading slash
* not clearing the original `/info/refs`

are indeed both bugs.

However, I'd caution you about this patch:

```c
+       if (pathlen > 0 && path[pathlen - 1] == '/')
+       {
+               path[pathlen - 1] = 0;
+       }
```

That's not a great pattern in general, and will definitely crash when `t->server.url.path == NULL` 
since you're stomping on a constant.

I think you want something more like printf("%s", strcmp(path, "/") ? path : "") in the request line creation.  It would be nice to avoid the `strcmp` but 🤷‍♂ .

I'm presently thinking about refactoring this code, so I'll keep this in mind, but taking a surgical fix from you seems reasonable while I'm doing that.

**Comment by josharian** [comment]
> That's not a great pattern in general, and will definitely crash when t->server.url.path == NULL
since you're stomping on a constant.

Ouch. Thanks. It has been a long time now since I wrote C in earnest. I'll send a PR shortly with a different approach.

> not clearing the original /info/refs

I took a second look at this. Some servers send the query back as part of the new Location during a redirect; others do not. The code as written works only for those that do not. If it would be helpful, I can whip up a small standalone Go http server that puts the query in the new Location. (Short version: Use github.com/gorilla/mux with net/http.)

It is not obvious to me where the fix belongs. Unless I hear otherwise, I will leave fixing this to you.
